sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c

```diff
@@ -0,0 +1,541 @@
+    ng_netif_hdr_t *netif_hdr;
+    ng_pktsnip_t *payload;
+
+    if (pkt == NULL) {
+        DEBUG("ng_netdev_eth: pkt was NULL");
+        return -EINVAL;
+    }
+
+    payload = pkt->next;
+
+    if (pkt->type != NG_NETTYPE_NETIF) {
+        DEBUG("ng_netdev_eth: First header was not generic netif header\n");
+        return -EBADMSG;
+    }
+
+    hdr->type = byteorder_htons(ng_nettype_to_ethertype(pkt->next->type));
+
+    netif_hdr = pkt->data;
+
+    /* set ethernet header */
+    if (netif_hdr->src_l2addr_len == NG_ETHERNET_ADDR_LEN) {
+        memcpy(hdr->dst, ng_netif_hdr_get_src_addr(netif_hdr),
+               netif_hdr->src_l2addr_len);
+    }
+    else {
+        dev_eth_t *ethdev = dev->ethdev;
+        ethdev->driver->get_mac_addr(ethdev, hdr->src);
+    }
+
+    if (netif_hdr->flags & NG_NETIF_HDR_FLAGS_BROADCAST) {
+        _addr_set_broadcast(hdr->dst);
+    }
+    else if (netif_hdr->flags & NG_NETIF_HDR_FLAGS_MULTICAST) {
+        _addr_set_multicast(hdr->dst, payload);
+    }
+    else if (netif_hdr->dst_l2addr_len == NG_ETHERNET_ADDR_LEN) {
+        memcpy(hdr->dst, ng_netif_hdr_get_dst_addr(netif_hdr),
+               NG_ETHERNET_ADDR_LEN);
+    }
+    else {
+        DEBUG("ng_netdev_eth: destination address had unexpected format\n");
+        return -EBADMSG;
+    }
+
+    DEBUG("ng_netdev_eth: send to %02x:%02x:%02x:%02x:%02x:%02x\n",
+          hdr->dst[0], hdr->dst[1], hdr->dst[2],
+          hdr->dst[3], hdr->dst[4], hdr->dst[5]);
+
+    data_len += sizeof(ng_ethernet_hdr_t);
+
+    while (payload != NULL) {
+        if ((data_len + payload->size) > NG_ETHERNET_MAX_LEN) {
+            DEBUG("ng_netdev_eth: Packet too big for ethernet frame\n");
+            return -ENOBUFS;
+        }
+
+        memcpy(send_buffer + data_len, payload->data, payload->size);
+
+        data_len += payload->size;
+        payload = payload->next;
+    }
+
+    /* Pad to minimum payload size.
+     * Linux does this on its own, but it doesn't hurt to do it here.
+     * As of now only tuntaposx needs this. */
+    if (data_len < (NG_ETHERNET_MIN_LEN)) {
+        DEBUG("ng_netdev_eth: padding data! (%d -> ", data_len);
+        memset(send_buffer + data_len, 0, NG_ETHERNET_MIN_LEN - data_len);
+        data_len = NG_ETHERNET_MIN_LEN;
+        DEBUG("%d)\n", data_len);
+    }
+
+    return data_len;
+}
+
+void dev_eth_isr(dev_eth_t* dev)
+{
+    (void)dev;
+    msg_t msg;
+
+    DEBUG("ng_netdev_eth: Trigger ISR event\n");
+
+    /* TODO: check whether this is an input or an output event
+       TODO: refactor this into general io-signal multiplexer */
+
+    msg.type = NG_NETDEV_MSG_TYPE_EVENT;
+    msg.content.value = _ISR_EVENT_RX;
+
+    if (msg_send(&msg, ng_netdev_eth.mac_pid) <= 0) {
+        puts("dev_eth_isr: possibly lost interrupt.");
+    }
+}
+
+void dev_eth_rx_handler(dev_eth_t* dev) {
+    (void)dev;
+    _rx_event(&ng_netdev_eth);
+}
+
+void dev_eth_linkstate_handler(dev_eth_t *dev, int newstate)
```

Comment:
These parameters are supposed to be used by the underlying network stack. I'd rather pass on the event than remove the parameters.